lib/mayaUsd/undo/OpUndoItems.h

```diff
@@ -0,0 +1,423 @@
+//
+// Copyright 2021 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef MAYAUSD_UNDO_OP_UNDO_ITEMS_H
+#define MAYAUSD_UNDO_OP_UNDO_ITEMS_H
+
+#include "OpUndoInfo.h"
+
+#include <mayaUsd/undo/UsdUndoableItem.h>
+
+#include <maya/MDGModifier.h>
+#include <maya/MDagModifier.h>
+#include <maya/MDagPath.h>
+#include <maya/MFnSet.h>
+#include <maya/MGlobal.h>
+#include <maya/MSelectionList.h>
+
+#include <memory>
+#include <vector>
+
+namespace MAYAUSD_NS_DEF {
+
+//------------------------------------------------------------------------------
+// NodeDeletionUndoItem
+//------------------------------------------------------------------------------
+
+/// \class NodeDeletionUndoItem
+/// \brief Record data needed to undo or redo a Maya DG sub-operation.
+///
+/// For node deletion, use the specialized undo item that tracks which objects
+/// have already been deleted and avoid double-deletions.
+class NodeDeletionUndoItem : public OpUndoItem
+{
+public:
+    /// \brief delete a node.
+    MAYAUSD_CORE_PUBLIC
+    static MStatus deleteNode(
+        const std::string name,
+        const MString&    nodeName,
+        const MObject&    node,
+        OpUndoInfo&       undoInfo);
+
+    /// \brief construct a Maya DG modifier recorder.
+    MAYAUSD_CORE_PUBLIC
+    NodeDeletionUndoItem(std::string name)
+        : OpUndoItem(std::move(name))
+    {
+    }
+
+    MAYAUSD_CORE_PUBLIC
+    ~NodeDeletionUndoItem() override;
```

Comment:
Looks like your destructor doesn't do anything beyond what the base class does.  If that's the case, you may want to just leave it out altogether and let the compiler generate it for you.  I guess if yu really want to declare it you could say
~NodeDeletionUndoItem() override = default;
but then what's the point of overriding it...

Reply:
I often override destructors so to be able to put breakpoint in them when needed.